src/sync.cpp

```diff
@@ -37,23 +39,26 @@ void PrintLockContention(const char* pszName, const char* pszFile, int nLine)
 //
 
 struct CLockLocation {
-    CLockLocation(const char* pszName, const char* pszFile, int nLine, bool fTryIn)
-    {
-        mutexName = pszName;
-        sourceFile = pszFile;
-        sourceLine = nLine;
-        fTry = fTryIn;
-    }
+    CLockLocation(
+        const char* pszName, const char* pszFile, int nLine, bool fTryIn, std::string thread_name_) :
+            fTry(fTryIn),
+            mutexName(pszName),
+            sourceFile(pszFile),
+            thread_name(std::move(thread_name_)),
+            sourceLine(nLine) { }
 
     std::string ToString() const
     {
-        return mutexName + "  " + sourceFile + ":" + itostr(sourceLine) + (fTry ? " (TRY)" : "");
+        return tfm::format(
+            "%s %s:%s%s (in thread %s)",
+            mutexName, sourceFile, itostr(sourceLine), (fTry ? " (TRY)" : ""), thread_name);
     }
 
 private:
     bool fTry;
     std::string mutexName;
     std::string sourceFile;
+    std::string thread_name;
```

Comment:
In commit "threads: add thread names to deadlock debugging message" (503c1ff6eeeb6e222ff446033745650ba59ab659)

In case you do take my earlier suggestion and change `GetInternalName` to return a const string reference instead of a temporary string, you could change this type to a const string reference and avoid any string copy initializing this struct field (it would point directly to the `g_thread_name` instance).

Reply:
In making this change I hit some compiler barf that has me out of my depths: https://gist.github.com/jamesob/5a19c75b34a49a4972b4105e0287d79d

Reply:
Oh, the relevant part of that error message is:

```
sync.cpp:133:33: note: in instantiation of member function 'std::vector<std::pair<void *, CLockLocation>, std::allocator<std::pair<void *, CLockLocation> > >::operator=' requested here
        lockdata.lockorders[p1] = g_lockstack;
```

It's complaining that it can't copy the g_lockstack variable, which makes sense since adding a reference member to CLockLocation makes it not copyable by default. If inclined, you could work around it by manually defining a CLockLocation copy constructor:

```
CLockLocation(const CLockLocation& other) : fTry(other.fTry), mutexName(other.mutexName), etc, ... {}
```

Reply:
I think I was wrong. Even with a reference member CLockLocation should still be copyable, just not assignable, so you may be able to fix this by emplacing into lockorders instead of assigning into it.

Reply:
Fixed by changing the above assignment to `emplace()` per your advice. Thanks!